### PR TITLE
removes empty filter parameters

### PR DESCRIPTION
**Pull request by yurii-github**
was
SOMEWEB/web/default/?BooksSearch[updated_date]=&BooksSearch[favorite]=&BooksSearch[title]=zzzzzzzzz&BooksSearch[isbn13]=&BooksSearch[read]=&BooksSearch[year]=&BooksSearch[ext]=

becomes
SOMEWEB/web/default/?BooksSearch[title]=zzzzzzzzz

### CODE REVIEW

File: framework/assets/yii.gridView.js [hunk 1]
```diff
@@ -101,6 +101,9 @@
             $grid.find('form.gridview-filter-form').remove();
             var $form = $('<form action="' + url + '" method="get" class="gridview-filter-form" style="display:none" data-pjax></form>').appendTo($grid);
             $.each(data, function (name, value) {
+                if (value == '') {
+            		return;
```
Comment (samdark) [inline framework/assets/yii.gridView.js]: Are you sure it should be `return` instead of `continue`?

Comment (yurii-github) [inline framework/assets/yii.gridView.js]: yeah. it is callback function, so it must end.

p.s.
not related to this topic -  still i have problem with removing model name from parameter for whole grid (i can remove it by setting custom filter per column one by one.. but that's not good, maybe you know how to do it in one place
i mean
[i]?BooksSearch[title]=zzzzzzzzz[/i]
to
[i]?title=zzzzzzzzz[/i]

Comment (samdark) [inline framework/assets/yii.gridView.js]: You could have 2 models used in a form so I'd not remove it.

**Comment by samdark** [comment]
Thanks!

**Comment by samdark** [comment]
Reverted since @qiangxue pointed out valid use case for previous code: https://github.com/yiisoft/yii2/commit/aba7710699706acb9828fa0a27a91b1d1ac0d46c

**Comment by SDKiller** [comment]
https://github.com/yiisoft/yii2/commit/aba7710699706acb9828fa0a27a91b1d1ac0d46c#commitcomment-10079019
> What if you have a pre-filled default filter value and you want to allow removing this filter value by providing an empty value?

Use case is valid, but but empty filter fields are more frequent, IMO.

Maybe add an option so developer could choose? 

Urls are ugly indeed.
In addition - if leave only not empty values url is more informative.

**Comment by samdark** [comment]
Ideally it should somehow check for defaults.

**Comment by yurii-github** [comment]
what is " pre-filled default filter value"? do you mean property defined in model class?
That's not good, but anyway, i have no idea how to workaround this usage for those who are using it (make it compatible).

i don't use extended model now as search model generated by Gii is nothing more than model + 1 filter/search function. which is a bit stupid, maybe it has some greater usage, not hit it so far.
 so my current usage of gridView looks like this now

    public function actionIndex()
    {
    	return $this->render('index', [
    		'searchModel' => ($filter = new Books()),
    		'dataProvider' => $filter->search(Yii::$app->request->queryParams)
    	]);
    }

and if someday it will require default filters set it wil look like

    public function actionIndex()
    {
    	return $this->render('index', [
    		'searchModel' => ($filter = new Books(['title' => 's'])),
    		'dataProvider' => $filter->search(Yii::$app->request->queryParams)
    	]);
    }


regards

**Comment by yurii-github** [comment]
i see what you mean. it resets..

**Comment by yurii-github** [comment]
on first try i was able to workaround it with custom filter by breaking html standart, dirty but it works

VIEW (grid config)
            $.each(data, function (name, value) {
            	if (value == '' && !$('[name="'+name+'"]').attr('can-be-empty')) {
            		return;
            	}

gridView,js - adding checker for attribute that can be empty

            $.each(data, function (name, value) {
            	if (value == '' && !$('[name="'+name+'"]').attr('can-be-empty')) {
            		return;
            	}

**Comment by RomeroMsk** [comment]
@yurii-github, I have the same issue and solving it like this for now:
```javascript
$(document).ready(function() {
    $("#search-form").on('submit', function() {
        $(this).find("[name]").each(function(index, elem) {
            if (!$(elem).val()) {
                $(elem).attr('name', null);
            }
        });
    });
});
```
And about your "p.s.": you are talking about `ActiveRecord::formName()`, I think.

**Comment by yurii-github** [comment]
RomeroMsk 
1. nice workaround with js.
2. thanks. it was really what i was hoping for to deface yii2

    public function formName() { return ''; }

perfect

**Pull request closed without merging**